packages/@lwc/engine-core/src/framework/components/dynamic.ts

```diff
@@ -0,0 +1,103 @@
+import { create, defineProperty } from '@lwc/shared';
+import { defaultEmptyTemplate } from '../secure-template';
+import { registerComponent } from '../component';
+import { LightningElement, LightningElementConstructor } from '../base-lightning-element';
+import { ComponentDef, getComponentInternalDef } from '../def';
+import {
+    appendVM,
+    computeShadowMode,
+    createVMContext,
+    getAssociatedVMIfPresent,
+    hasWireAdapters,
+    resetComponentRoot,
+    runConnectedCallback,
+    runDisconnectedCallback,
+    VM,
+    VMState,
+} from '../vm';
+import { invokeComponentConstructor } from '../invoker';
+import { installWireAdapters } from '../wiring';
+
+function replaceDefinitionForVM(vm: VM, newDef: ComponentDef, replacedCallback: () => void) {
+    if (vm.state === VMState.connected) {
+        runDisconnectedCallback(vm);
```

Comment:
Assuming we enable native connected/disconnected callback events (#3198), this would be the one place where we are simulating our own disconnected callback behavior. This introduces unknown unknowns – e.g. subtle bugs where we incorrectly detect connectedness in templates (like https://github.com/salesforce/lwc/issues/2609#issuecomment-1055826907).

Are we sure we want to add another source of synthetic lifecycle callbacks? Is there some other way we can achieve this? (E.g. explicitly removing and re-adding the DOM node.)

Reply:
Hmm, I don't understand this part. How is this different from rehydration from SSR?

Reply:
Rehydration is another case where we manually invoke connected callback, you're right. Although I'm not sure we will always do it that way. /cc @divmain 

In Lit for example I believe they use the native `connectedCallback`, but they queue up the callbacks in SSR so that they fire in a predictable order. (Otherwise they will fire based on when `customElements.define()` is called, which is kind of random and unpredictable.

Reply:
A compelling reason hasn't surfaced yet to move us away from manually invoking `connectedCallback` during rehydration. Lit uses the pattern you mention, as well as adding/removing ephemeral attrs (with corresponding event listeners) to enforce deterministic order of execution through a component hierarchy. The reason we were originally considered moving in this direction has gone away, now that we've decided to lock down DOM APIs on the host element.